File: android/feature/client/src/main/kotlin/com/devora/mencare/feature/client/appointments/AppointmentsViewModel.kt
package com.devora.mencare.feature.client.appointments

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.devora.mencare.core.common.AppError
import com.devora.mencare.core.data.network.DataSync
import com.devora.mencare.core.data.network.SyncKeys
import com.devora.mencare.core.data.repository.AuthRepository
import com.devora.mencare.core.data.repository.BookingRepository
import com.devora.mencare.core.data.repository.CatalogRepository
import com.devora.mencare.core.model.Appointment
import com.devora.mencare.core.model.AppointmentStatus
import com.devora.mencare.core.model.CancellationActor
import com.devora.mencare.core.model.Operator
import com.devora.mencare.core.model.Service
import com.devora.mencare.core.model.WaitlistEntry
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import java.time.LocalDateTime
import java.time.temporal.ChronoUnit
import javax.inject.Inject

data class AppointmentsUiState(
    val upcoming: List<Appointment> = emptyList(),
    val past: List<Appointment> = emptyList(),
    val waitlist: List<WaitlistEntry> = emptyList(),
    val services: Map<String, Service> = emptyMap(),
    val operators: Map<String, Operator> = emptyMap(),
    val totalVisits: Int = 0,
    /** Nome di chi l'ha servito più spesso. Al posto della spesa: le cifre le vede solo il titolare. */
    val favoriteOperatorName: String? = null,
    val avgDaysBetweenVisits: Int? = null,
    val lastCompleted: Appointment? = null,
    val loading: Boolean = false,
    val error: AppError? = null,
) {
    /** Prima risposta non ancora arrivata: non c'è ancora niente da elencare. */
    val isEmpty: Boolean
        get() = upcoming.isEmpty() && past.isEmpty() && waitlist.isEmpty() && services.isEmpty()
}

@OptIn(ExperimentalCoroutinesApi::class)
@HiltViewModel
class AppointmentsViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val bookingRepository: BookingRepository,
    catalogRepository: CatalogRepository,
    private val sync: DataSync,
) : ViewModel() {

    private val clientAppointments = authRepository.currentUser.flatMapLatest { user ->
        val id = user?.clientRecordId
        if (id == null) flowOf(emptyList()) else bookingRepository.appointmentsForClient(id)
    }

    private val clientWaitlist = authRepository.currentUser.flatMapLatest { user ->
        val id = user?.clientRecordId
        if (id == null) flowOf(emptyList()) else bookingRepository.waitlistForClient(id)
    }

    val state = combine(
        clientAppointments,
        clientWaitlist,
        catalogRepository.services,
        catalogRepository.operators,
        sync.state(SyncKeys.APPOINTMENTS, SyncKeys.WAITLIST, SyncKeys.CATALOG),
    ) { appointments, waitlist, services, operators, status ->
        val now = LocalDateTime.now()
        val completed = appointments
            .filter { it.status == AppointmentStatus.COMPLETED }
            .sortedBy { it.start }
        val visitDates = completed.map { it.date }.distinct()
        AppointmentsUiState(
            upcoming = appointments
                .filter { it.isActive && it.end >= now }
                .sortedBy { it.start },
            past = appointments
                .filter { !it.isActive || it.end < now }
                .sortedByDescending { it.start },
            waitlist = waitlist,
            services = services.associateBy { it.id },
            operators = operators.associateBy { it.id },
            totalVisits = completed.size,
            favoriteOperatorName = completed
                .groupingBy { it.operatorId }.eachCount()
                .maxByOrNull { it.value }
                ?.let { operators.firstOrNull { op -> op.id == it.key } }
                ?.name?.substringBefore(' '),
            avgDaysBetweenVisits = visitDates
                .zipWithNext { a, b -> ChronoUnit.DAYS.between(a, b) }
                .takeIf { it.isNotEmpty() }
                ?.average()?.toInt(),
            lastCompleted = completed.lastOrNull(),
            loading = status.loading,
            error = status.error,
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), AppointmentsUiState())

    /** "Riprova": si dimenticano i guasti e si rilegge tutto ciò che è osservato. */
    fun retry() = sync.retry()

    fun cancel(appointmentId: String) {
        viewModelScope.launch {
            bookingRepository.cancel(appointmentId, CancellationActor.CLIENT)
        }
    }

    fun leaveWaitlist(entryId: String) {
        viewModelScope.launch {
            bookingRepository.leaveWaitlist(entryId)
        }
    }
}

File: ios/MenCare/Features/Client/Appointments/AppointmentsViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class AppointmentsViewModel {

    private let auth: AuthRepository
    private let booking: BookingRepository
    private let catalog: CatalogRepository

    let state = LoadState()

    private(set) var upcoming: [Appointment] = []
    private(set) var past: [Appointment] = []
    private(set) var waitlist: [WaitlistEntry] = []
    private(set) var services: [String: Service] = [:]
    private(set) var operators: [String: Operator] = [:]

    private(set) var totalVisits = 0
    /// Nome di chi l'ha servito piu' spesso. Al posto della spesa: le cifre le
    /// vede solo il titolare.
    private(set) var favoriteOperatorName: String?
    private(set) var avgDaysBetweenVisits: Int?
    private(set) var lastCompleted: Appointment?

    /// Un annullamento o un'uscita dalla coda che non riesce va detto: prima
    /// spariva in silenzio e la riga restava dov'era senza spiegazioni.
    var actionError: String?

    init(auth: AuthRepository, booking: BookingRepository, catalog: CatalogRepository) {
        self.auth = auth
        self.booking = booking
        self.catalog = catalog
    }

    func load() async {
        await state.run {
            let catalogSnapshot = try await catalog.catalog()
            let appointments = try await booking.myAppointments()
            let queue = try await booking.myWaitlist()

            services = Dictionary(uniqueKeysWithValues: catalogSnapshot.services.map { ($0.id, $0) })
            operators = Dictionary(uniqueKeysWithValues: catalogSnapshot.operators.map { ($0.id, $0) })
            waitlist = queue

            let now = LocalDateTime.now()
            upcoming = appointments.filter { $0.isActive && $0.end >= now }.sorted { $0.start < $1.start }
            past = appointments.filter { !$0.isActive || $0.end < now }.sorted { $0.start > $1.start }

            let completed = appointments.filter { $0.status == .completed }.sorted { $0.start < $1.start }
            totalVisits = completed.count
            lastCompleted = completed.last
            favoriteOperatorName = favourite(in: completed)
            avgDaysBetweenVisits = cadence(of: completed)
        }
    }

    private func favourite(in completed: [Appointment]) -> String? {
        let counts = Dictionary(grouping: completed, by: \.operatorId).mapValues(\.count)
        guard let top = counts.max(by: { $0.value < $1.value })?.key else { return nil }
        return operators[top]?.name.split(separator: " ").first.map(String.init)
    }

    private func cadence(of completed: [Appointment]) -> Int? {
        let dates = completed.map(\.date).reduce(into: [LocalDate]()) { acc, d in
            if acc.last != d { acc.append(d) }
        }
        guard dates.count > 1 else { return nil }
        let gaps = zip(dates, dates.dropFirst()).map { $0.daysBetween($1) }
        return gaps.reduce(0, +) / gaps.count
    }

    func cancel(_ appointmentId: String) {
        Task {
            actionError = nil
            switch await booking.cancel(appointmentId, by: .client) {
            case .success: await load()
            case .failure(let error): actionError = error.displayMessage
            }
        }
    }

    func leaveWaitlist(_ entryId: String) {
        Task {
            actionError = nil
            switch await booking.leaveWaitlist(entryId) {
            case .success: await load()
            case .failure(let error): actionError = error.displayMessage
            }
        }
    }
}
